runtime/tests/checker/type_inference_test.go

```diff
@@ -0,0 +1,293 @@
+
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+func TestArrayElementTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("numeric array", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [Int8] = [1, 2, 3]
+				var y: [Int8]? = [1, 2, 3]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("anystruct array", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [AnyStruct] = [1, 2, 3]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("invalid", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [Int8]? = [1, 534, 3]
+			}
+		`)
+
+		require.Error(t, err)
+		assert.IsType(t, &sema.CheckerError{}, err)
+	})
+
+	t.Run("anystruct", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: AnyStruct = [1, 534534, 3]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("inferring from rhs", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x = [1, 534534, 3]
+				var y: Int = x[2]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("nested array", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: [[[Int8]]] = [[[1, 2, 3], [4]], []]
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+}
+
+func TestDictionaryTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("numerics", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int8:Int64} = {0: 6, 1: 5}
+				var y: {Int8:Int64?} = {0: 6, 1: 5}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("heterogeneous", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:AnyStruct} = {0: 6, 1: "hello", 2: nil}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("nested", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:{Int:{Int:AnyStruct}}} = {0: {0: {0: 6}, 1: {0: 7}}}
+			}
+		`)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("invalid", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int8:Int64} = {23423:6, 1:5}
+			}
+		`)
+
+		require.Error(t, err)
+		assert.IsType(t, &sema.CheckerError{}, err)
+	})
+
+	t.Run("nested invalid", func(t *testing.T) {
+		_, err := ParseAndCheck(t, `
+			fun test() {
+				var x: {Int:{Int:{Int:Int8}}} = {0: {0: {0: "hello"}, 1: {0: 7}}}
+			}
+		`)
+
+		require.Error(t, err)
+		assert.IsType(t, &sema.CheckerError{}, err)
+	})
+}
+
+func TestReturnTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		fun test(): [Int8] {
+			return [1, 2, 3]
+		}
+	`)
+
+	require.NoError(t, err)
+}
+
+func TestFunctionArgumentTypeInference(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+		fun test() {
+			foo(a: [1, 2, 3])
+		}
+
+		fun foo(a: [Int8]) {
+		}
+	`)
+
+	// Type inferring for function arguments is not supported yet.
+	require.Error(t, err)
+	assert.IsType(t, &sema.CheckerError{}, err)
```

Comment:
Could you please make assertion more specific here, i.e. there is one error, a type mismatch?

I verified this is also the current behaviour 👍